zqd/api/connection.go

```diff
@@ -0,0 +1,261 @@
+	// ErrSpaceNotFound returns when specified space does not exist.
+	ErrSpaceNotFound = errors.New("space not found")
+	// ErrSpaceExists returns when specified the space already exists.
+	ErrSpaceExists = errors.New("space exists")
+)
+
+type Connection struct {
+	client *resty.Client
+}
+
+func newConnection(useragent string, client *resty.Client) *Connection {
+	if useragent != "" {
+		client.SetHeader("User-Agent", useragent)
+	}
+	client.SetError(Error{})
+	client.OnAfterResponse(checkError)
+	return &Connection{
+		client: client,
+	}
+}
+
+// NewConnection creates a new connection with the given useragent string
+// and a base URL set up to talk to http://localhost:defaultport
+func NewConnection(useragent string) *Connection {
+	u, _ := url.Parse("http://localhost:" + strconv.Itoa(DefaultPort))
+	return NewConnectionToURL(useragent, u)
+}
+
+// NewConnectionTo creates a new connection with the given useragent string
+// and a base URL derived from the hostURL argument.
+func NewConnectionTo(useragent string, hostURL string) (*Connection, error) {
+	url, err := url.Parse(hostURL)
+	if err != nil {
+		return nil, err
+	}
+	return NewConnectionToURL(useragent, url), nil
+}
+
+// NewConnectionToURL creates a new connection object with the given useragent string
+// and a base URL specified by the url argument.
+func NewConnectionToURL(useragent string, url *url.URL) *Connection {
+	client := resty.New()
+	client.HostURL = url.String()
+	c := newConnection(useragent, client)
+	return c
+}
+
+func (c *Connection) Do(ctx context.Context, method, url string, body interface{}) (*resty.Response, error) {
+	req := c.Request(ctx).SetBody(body)
+	return req.Execute(method, url)
+}
+
+func checkError(client *resty.Client, resp *resty.Response) error {
+	if resp.IsSuccess() {
+		return nil
+	}
+	resErr := &ErrorResponse{Response: resp}
+	if err := resp.Error(); err != nil {
+		resErr.Err = err.(*Error)
+	} else {
+		resErr.Err = errors.New(resp.String())
+	}
+	return resErr
+}
+
+func (c *Connection) stream(req *resty.Request) (io.ReadCloser, error) {
+	resp, err := req.SetDoNotParseResponse(true).Send() // disables middleware
+	if err != nil {
+		return nil, err
+	}
+	r := resp.RawBody()
+	if resp.IsSuccess() {
+		return r, nil
+	}
+	defer r.Close()
+	body, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	resErr := &ErrorResponse{Response: resp}
+	if resty.IsJSONType(resp.Header().Get("Content-Type")) {
+		var apierr *Error
+		if err := json.Unmarshal(body, apierr); err != nil {
+			return nil, err
+		}
+		resErr.Err = apierr
+	} else {
+		resErr.Err = errors.New(string(body))
+	}
+	return nil, resErr
+}
+
+// SetTimeout sets the underlying http request timeout to the given duration
+func (c *Connection) SetTimeout(to time.Duration) {
+	c.client.SetTimeout(to)
+}
+
+func (c *Connection) URL() string {
+	return c.client.HostURL
+}
+
+func (c *Connection) SetURL(u string) {
+	c.client.SetHostURL(u)
+}
+
+type ErrorResponse struct {
```

Comment:
Nit: Maybe move this and its methods to the bottom of the file so it isn't in the middle of the Connection methods.

Reply:
sure